### PR TITLE
Feature: Comply with framework registration spec

### LINKED ISSUE #183: Remove Fields by Default
If you have the `website: true` option enabled, and nothing else, the registration page looks like this:

screenshot 2015-11-02 10 36 08

As you can see, there are lots of fields there. Previously, we only showed the:

- first name
- last name
- email
- password

Fields, which is what we should return to doing. This way users (by default) aren't overwhelmed with all of the fields available.

### PR DESCRIPTION
**Pull request by typerandom**
Complies with the [framework spec PR](https://github.com/stormpath/stormpath-framework-spec/pull/15) by removing certain registration fields from the config file and patching places where these fields were used.

Closes #183

### CODE REVIEW

File: lib/controllers/register.js [hunk 1]
```diff
@@ -81,11 +81,13 @@ function defaultJsonResponse(req, res) {
  */
 
 function applyDefaultAccountFields(config, req) {
-  if (!config.web.register.fields.givenName.enabled && !req.body.givenName) {
+  var registerFields = config.web.register.fields;
+
+  if ((registerFields.givenName && !registerFields.givenName.enabled) && !req.body.givenName) {
```
Comment (timothyej) [inline lib/controllers/register.js]: Why the extra parentheses here? :)

File: lib/controllers/register.js [hunk 2]
```diff
@@ -81,11 +81,13 @@ function defaultJsonResponse(req, res) {
...
     req.body.givenName = 'Anonymous';
   }
 
-  if (!config.web.register.fields.surname.enabled && !req.body.surname) {
+  if ((registerFields.surname && !registerFields.surname.enabled) && !req.body.surname) {
```
Comment (timothyej) [inline lib/controllers/register.js]: And why extra parentheses here too?

File: lib/controllers/register.js (hunk 1 above)
Comment (typerandom) [inline lib/controllers/register.js]: Fixed:)

**Comment by robertjd** [comment]
This introduces an `enabled` property, which is not something that we have in the spec for registration fields.  My understanding is that a field is either an object or null.  

If it is null, it is not shown.

If it is an object, we show the field.

If the object has the `required` property, we use form validation to force the user to provide the value.

For the `givenName` and `surname` fileds, if `required` is `false`, we populate the value 'UNKNOWN' before we submit the account data to the API (because the API requires it).  Note: we are currently using 'Anonymous', but I do believe 'UNKNOWN` is better and this is what we've defined in the framework spec.

Ping @rdegges on all this ^

**Comment by rdegges** [comment]
Agree w/ Robert -- this is also the old functionality.

**Comment by robertjd** [comment]
Update: after discussing it further, we decided to go the route of using the `enabled` property because trying to extend layers of configuration, where you can define a field as null, would be madness.  Also, with the enabled strategy we can define all of our default supported fields in our config.yaml, but just disable the ones that we don't want to render by default.

This PR has been updated accordingly.

File: docs/changelog.rst [hunk 3]
```diff
@@ -6,11 +6,30 @@ Change Log
 
 All library changes, in descending order.
 
-Version 2.2.1
+Version 2.3.0
 --------------
 
 **Not Yet Released**
 
+May fixes for the registration field configuration:
```
Comment (rdegges) [inline docs/changelog.rst]: *Many

File: docs/changelog.rst [hunk 4]
```diff
@@ -6,11 +6,30 @@ Change Log
...
+
+- Custom fields, as defined in the ``register`` block, will now appear in our
+  default registration form (they were not appearing before).
+
+- Custom fields now *must* be defined in the ``register`` block, otherwise the
+  data will be rejected during account creation.
+
+- Only the First Name, Last Name, Email, and Password fields are shown by
+  default (the middle name and username fields are no longer shown be default).
```
Comment (rdegges) [inline docs/changelog.rst]: *by

File: docs/registration.rst [hunk 5]
```diff
@@ -38,40 +38,95 @@ we will cover them in detail below:
...
-* lastName
+* givenName
+* surname
 * email
 * password
 
-While email and password will always be required (you'll get an API error if
-you omit them), you may not need to require first and last name.  These
-can be configured, and we'll cover that in the next section.
+While email and password will always be required, you may not need to require
```
Comment (rdegges) [inline docs/registration.rst]: *may not = do not

File: lib/config.yml [hunk 6]
```diff
@@ -27,51 +27,51 @@ web:
...
         placeholder: "Password"
         required: true
         type: "password"
-      passwordConfirm:
-        name: "passwordConfirm"
-        placeholder: "Confirm Password"
-        required: false
-        type: "password"
     fieldOrder:
-      - "username"
```
Comment (rdegges) [inline lib/config.yml]: I think username should be listed first. Above givenName / surname. It makes more sense this way in most registration pages.

File: lib/config.yml [hunk 7]
```diff
@@ -27,51 +27,51 @@ web:
     enabled: false
     uri: "/register"
     nextUri: "/"
-    autoAuthorize: false
+    autoLogin: false
```
Comment (rdegges) [inline lib/config.yml]: :( Can't we just set to true anyway? I know we can't, but it's sooo tempting.

**Comment by rdegges** [comment]
Made some comments -- but it looks good to me. +1

**Comment by robertjd** [comment]
*boosh* FIX EVERYTHING

thanks @rdegges for all the feedback!

**Comment by typerandom** [comment]
@robertjd Awesome!

**Pull request merged**